### PR TITLE
Deprecate split_non_commuting

### PR DESCRIPTION
**Pull request by AlbertMitjans**
1. Deprecate `split_non_commuting`. Use `split_tape` instead.

### CODE REVIEW

File: pennylane/_device.py [hunk 1]
```diff
@@ -725,51 +712,8 @@ def batch_transform(self, circuit: QuantumScript):
...
-                not isinstance(m, (SampleMP, ProbabilityMP, CountsMP)) for m in circuit.measurements
-            )
-        ):
-            # Check for case of non-commuting terms and that there are no Hamiltonians
-            # TODO: allow for Hamiltonians in list of observables as well.
-            circuits, hamiltonian_fn = qml.transforms.split_non_commuting(circuit)
-
-        else:
-            # otherwise, return the output of an identity transform
-            circuits = [circuit]
```
Comment (rmoyard) [inline pennylane/_device.py:769]: Do we still cover the case where there is no split?

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -892,19 +892,9 @@ def statistics(
                 )
 
             elif obs.return_type is Shadow:
-                if len(measurements) > 1:
```
Comment (rmoyard) [inline pennylane/_qubit_device.py:895]: That's great :+1:

File: pennylane/transforms/split_tape.py [hunk 3]
```diff
@@ -200,40 +202,42 @@ def split_tape(tape: QuantumTape, group=True):
...
         else:
-            coeff = None
             if isinstance(obs, SProd) and isinstance(m, ExpectationMP):
                 coeff = obs.scalar
                 m = qml.expval(obs.base)
 
             if m.hash not in measurements_dict:
                 measurements_dict[m.hash] = m
-                idxs_coeffs_dict[m.hash] = [(idx, coeff)]
+                idxs_coeffs_dict[m.hash] = [(idx, None)]
```
Comment (rmoyard) [inline pennylane/transforms/split_tape.py:211]: Why this change is necessary?

File: pennylane/transforms/split_tape.py [hunk 4]
```diff
@@ -264,6 +268,10 @@ def processing_fn(expanded_results):
         # sort results by idx
         results = tuple(res_dict[key] for key in sorted(res_dict))
 
+        if qml.math.requires_grad(expanded_results[0]):
+            # when computing gradients, we need to convert the tuple to a gradient box
+            return qml.math.stack(results)
```
Comment (rmoyard) [inline pennylane/transforms/split_tape.py:273]: The stacking should be done on the user side for the new return type, I think you just need to add `and not qml.active_return()`

File: tests/devices/test_default_qubit_broadcasting.py [hunk 5]
```diff
@@ -2023,7 +2025,7 @@ def circuit():
 
         circuit()
         # evaluated one expval altogether
-        assert spy.call_count == 1
+        assert spy.call_count == 2
```
Comment (rmoyard) [inline tests/devices/test_default_qubit_broadcasting.py:2028]: Can you explain what made the behaviour change here? I guess now the tape is split, but it was not the case before?

File: tests/shadow/test_shadow_expval.py [hunk 6]
```diff
@@ -118,20 +118,22 @@ def test_shots_none_error(self):
...
             return qml.shadow_expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(0))
 
-        msg = "Classical shadows cannot be returned in combination with other return types"
-        with pytest.raises(qml.QuantumFunctionError, match=msg):
-            shadow = circuit()
+        res = circuit()
+
+        assert len(res) == 2
+        assert res[0].shape == ()
+        assert res[1].shape == ()
```
Comment (rmoyard) [inline tests/shadow/test_shadow_expval.py:136]: Could you also test that the expval is correct?

File: tests/transforms/test_split_non_commuting.py [hunk 7]
```diff
@@ -59,7 +60,6 @@ def test_commuting_group_no_split(self, mocker):
             qml.Hadamard(0)
             qml.CNOT((0, 1))
             qml.expval(qml.PauliZ(0))
-            qml.expval(qml.PauliZ(0))
```
Comment (rmoyard) [inline tests/transforms/test_split_non_commuting.py:62]: Is it necessary to be removed ?

File: tests/transforms/test_split_non_commuting.py [hunk 8]
```diff
@@ -128,16 +128,6 @@ def test_different_measurement_types(self, meas_type):
             for meas in new_tape.measurements:
                 assert meas.return_type == the_return_type
 
-    def test_raise_not_supported(self):
-        """Test that NotImplementedError is raised when probabilities or samples are called"""
-        with qml.queuing.AnnotatedQueue() as q:
-            qml.expval(qml.PauliZ(0))
```
Comment (rmoyard) [inline tests/transforms/test_split_non_commuting.py:134]: Could you add the test that shows the correct results for this new behaviour?

**Review by rmoyard — COMMENTED** [review]
Thanks @AlbertMitjans! :100: Some initial comments and questions

File: pennylane/transforms/split_tape.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/transforms/split_tape.py:211]: It's not necessary! But a small code improvement (`coeff` was set to `None` before the `SProd` check).

File: tests/transforms/test_split_non_commuting.py (hunk 7 above)
Comment (AlbertMitjans) [reply tests/transforms/test_split_non_commuting.py:62]: Yes, because when having two equal measurements, `split_tape` will only execute one and copy the result. Thus the tape object won't be the same (which is what we are testing here).
